### PR TITLE
Fixes applied to migration process

### PR DESCRIPTION
**Pull request by emmanuelm41**
It solves issues reported on https://github.com/drand/drand/issues/855

### CODE REVIEW

**Comment by hsanjuan** [comment]
It's probably best practice (and likely faster), to wrap the reader/writers with `bufio`. Also, perhaps it would be more solid to do `os.Rename` and then copy back the file into the original location, and if that fails, perhaps rename the files back.

In one case, the node can go on without the backup, even on error when creating it. On the other, it cannot go on. One likely scenario might be a node with little disk space left, that cannot duplicate the DB. It will fail again and again to upgrade with the current approach. When renaming, it will upgrade just fine and could potentially continue.

**Review by nikkolasg — APPROVED** [review]
I'm personally fine for this change - the DB should not be THAT big, the problem is more out of memory error, which is a recurrent drand issue about go routine leakage I think.
So using `io.Copy` with streams is good for me.

File: fs/fs.go [hunk 1]
```diff
@@ -137,16 +138,28 @@ func FolderExists(folderPath, name string) bool {
...
+	if dest, err = os.Create(destFilePath); err != nil {
 		return err
 	}
+	defer dest.Close()
 
-	err = os.WriteFile(destFilePath, input, rwFilePermission)
+	_, err = io.Copy(dest, src)
 	if err != nil {
 		return err
 	}
```
Comment (nikkolasg) [inline fs/fs.go:158]: Actually can you add `err = dest.Sync()` to be sure the copied file is not corrupted ?

**Comment by nikkolasg** [comment]
What do you think about just using the "cp" cmd actually instead of re
implementing cp again ? Letting the OS do it is probably the easiest and
fastest way no ?

**Comment by emmanuelm41** [comment]
> It's probably best practice (and likely faster), to wrap the reader/writers with `bufio`. Also, perhaps it would be more solid to do `os.Rename` and then copy back the file into the original location, and if that fails, perhaps rename the files back.
> 
> In one case, the node can go on without the backup, even on error when creating it. On the other, it cannot go on. One likely scenario might be a node with little disk space left, that cannot duplicate the DB. It will fail again and again to upgrade with the current approach. When renaming, it will upgrade just fine and could potentially continue.

The original files are left untouched as the process must allow to execute a rollback if something goes wrong. That is something the team wanted. I try not to touch the original files so the risk is minimum. I personally think that is a good approach in this case. 

Talking about `bufio`, I can add it to the migration process. 

> What do you think about just using the "cp" cmd actually instead of re
> implementing cp again ? Letting the OS do it is probably the easiest and
> fastest way no ?

I am not sure if the control we have over the copy process is better or not. I feel confortable to copy files from golang

File: fs/fs.go (hunk 1 above)
Comment (emmanuelm41) [reply fs/fs.go:195]: Now we force to flush data, so it is not necessary anymore!

**Comment by hsanjuan** [comment]
> What do you think about just using the "cp" cmd actually instead of re
> implementing cp again ? Letting the OS do it is probably the easiest and
> fastest way no ?

No, we don't call the OS for these things. Nothing tells you that there is a shell to invoke, a cp command present in the system etc. In general invoking shell commands is going to be a bad thing in 99% of cases, if there is an alternative.

**Comment by emmanuelm41** [comment]
@hsanjuan could you please approve it if you see it is correct now? 😄 Thanks!

**Comment by nikkolasg** [comment]
Fine if we want to use buffered io - in this use case I dont see it as a requirement since we are reading from a local file so we already have all the data to fill the buffer, not like TCP socket etc-  But I am not fine to re-implement things that golang already provides. The [copy method from Go](https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/io/io.go;l=401;drc=refs%2Ftags%2Fgo1.17.3) is much more fine grained tested and we should use that.
If we want to control the size of the buffer (which is the only detail that differ between using `io.Copy` and buff writer from what I see), can we use native features such as https://pkg.go.dev/io#CopyBuffer ? Feels like it shoudl fulfill the job, right ?

**Comment by emmanuelm41** [comment]
@nikkolasg I just applied the change you suggested. What do you think now?

**Review by willscott — APPROVED** [review]

**Review by nikkolasg — APPROVED** [review]

File: fs/fs.go [hunk 2]
```diff
@@ -2,14 +2,17 @@
...
 	"fmt"
+	"io"
 	"os"
 	"os/user"
 	"path"
 )
 
 const defaultDirectoryPermission = 0740
 const rwFilePermission = 0600
+const copyChunkSize = 1024
```
Comment (hsanjuan) [inline fs/fs.go:15]: The main point of buffering is to not hammer the disk on read/write. This is even for the case when your source or dest actually refuse to read as much as you request. Do you know how reading a File works when the Filesystem is actually EBS and network-based? Does it return all bytes asked? Does go handle that for us? I have no idea. But that is the point.

Now CopyBuffer works I guess to hammer the disk less, assuming the file reader/writers do behave properly.

Now, if you are going to manually specify the buffer, then don't make it worse than the default. which is 32KB and not 1KB:

https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/io/io.go;l=412

I was hoping we could copy in something larger than 32KB chunks. We should definitely not be copying in 1KB chunks (!).

Sorry if it sounds pendantic, but my problem is that there seems to be 0 attention to these things. Did I not say at the beginning that I'd prefer if we write in larger chunks? What am I supposed to say when the result is we now want to write in chunks 32 times smaller?

**Review by hsanjuan — CHANGES_REQUESTED** [review]

File: fs/fs.go [hunk 3]
```diff
@@ -137,13 +140,60 @@ func FolderExists(folderPath, name string) bool {
...
+	if src, err = os.Open(origFilePath); err != nil {
 		return err
 	}
 
-	err = os.WriteFile(destFilePath, input, rwFilePermission)
-	if err != nil {
+	// close fi on exit and check for its returned error
+	defer func() {
+		if err := src.Close(); err != nil {
+			panic(err)
```
[the remark this answers is not in the document]
Comment (hsanjuan) [reply fs/fs.go:153]: Or ignore the error... certainly those options are better than having a panic. What is a user meant to make out of this panic if it happens?

[the remark this answers is not in the document]
Comment (jleni) [reply fs/fs.go:153]: I would at least log that there was a problem when trying to close...
I guess there is some small risk is that data is not flushed properly and there is some corruption that goes unnoticed?

File: fs/fs.go (hunk 2 above)
Comment (emmanuelm41) [reply fs/fs.go]: Hi @hsanjuan, I have updated the value to 128KB. What do you think now?

File: fs/fs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply fs/fs.go]: I think it is a good point to handle the error on these deferred functions. Now we will rollback the migration if something fails.

**Review by hsanjuan — APPROVED** [review]

**Pull request merged**